File: Patterns/Observer/php/www/subject/WeatherProvider.php
<?php
	require_once 'ISubject.php';
	require_once 'WeatherVO.php';
	
	class WeatherProvider implements ISubject{
		
		private $observers = array();
		private $weatherVO;
		
		public function __construct(){
			$this->weatherVO = new WeatherVO;
		}
		
		public function registerObserver(IObserver $observer){
			
			foreach ($this->observers as $observerIndex => $registeredObserver) {
				if($observer === $registeredObserver)
					return;
			}
			
			array_push($this->observers, $observer);
				
		}
		
		public function removeObserver(IObserver $observer){
			foreach ($this->observers as $observerIndex => $registeredObserver) {
				if($observer == $registeredObserver)
					unset($this->observers[$observerIndex]);
					break;
			}
		}
		
		public function notifyObservers(){
			foreach ($this->observers as $observerIndex => $registeredObserver) {
				$registeredObserver->update($this->weatherVO);
			}
		}
		
		public function setWeatherData($temperature, $humidity, $pressure){
			$this->weatherVO->temperature = $temperature;
			$this->weatherVO->humidity = $humidity;
			$this->weatherVO->pressure = $pressure;
			$this->notifyObservers();
		}
		
	}

File: Patterns/Observer/java/src/subject/WeatherProvider.java
package subject;

import java.util.ArrayList;

import observer.IObserver;

public class WeatherProvider implements ISubject {
	private ArrayList<IObserver> observers = new ArrayList<IObserver>();
	private WeatherVO weatherVO = new WeatherVO();
	
	
	
	public WeatherProvider() {
		
	}

	@Override
	public void registerObserver(IObserver observer) {
		if(!observers.contains(observer)){
			observers.add(observer);
		}
	}

	@Override
	public void removeObserver(IObserver observer) {
		if(observers.contains(observer)){
			observers.remove(observer);
		}
	}

	@Override
	public void notifyObservers() {
		for (IObserver observer : observers) {
			observer.update(weatherVO);
		}

	}
	
	public void setWeatherData(float temperature, float humidity, float pressure){
		weatherVO.temperature = temperature;
		weatherVO.humidity = humidity;
		weatherVO.pressure = pressure;
		notifyObservers();
	}

}
